[Mailer] [Smtp] Add DSN param to enforce TLS/STARTTLS

ssddanbrown opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 7.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | Fix #48297
| License       | MIT


Adds 'require_tls' param which can be set to true to enforce the use of TLS/STARTTLS within the ESMTP transport.
This was discussed in #48297.
These changes are based upon patches [I've been maintaining](https://github.com/ssddanbrown/symfony-mailer/commit/e9de8dccd76a63fc23475016e6574da6f5f12a2d) for my own projects.

This is my first PR to Symfony, I've tried to follow the guide as best as possible, and I was also using #53621 as a general guide. There are some other ways I could have gone about things, but I've tried to avoid touching as much existing Symfony code as possible. 

In #48297, nicolas-grekas mentioned unifying such an option with `auto_tls` under a `tls` option, but I think these are distinct options which may not be as clear combined (in addition to any expectations of such an option disabling/enabling TLS in general).

mtarld commented on src/Symfony/Component/Mailer/Transport/Smtp/EsmtpTransportFactory.php:
```diff
@@ -75,6 +75,10 @@ public function create(Dsn $dsn): TransportInterface
             $transport->setPingThreshold((int) $pingThreshold);
         }
 
+        if (null !== ($tlsRequired = $dsn->getOption('require_tls'))) {
```
You can leverage the new `getBoolOption`

mtarld commented on src/Symfony/Component/Mailer/Tests/Transport/Smtp/EsmtpTransportTest.php:
```diff
@@ -297,6 +297,25 @@ public function testSocketTimeout()
+        $message->addTo('recipient@example.org');
+        $message->text('.');
+
+        try {
```
Prefer using `expectException` and `exceptExceptionMessage`.

ssddanbrown committed:
[Mailer] [Smtp] Add DSN param to enforce TLS/STARTTLS

Adds 'require_tls' param which can be set to true to enforce
the use of TLS/STARTTLS within the ESMTP transport.

ssddanbrown commented:
Thanks for the review @mtarld.
I've now addressed both of the raised points, and have re-based the branch to the current 7.3 state.

nicolas-grekas reviewed (approved):
LGTM.

We need to make it clear in the doc that this option makes sense with the "smtp" scheme (since TLS is always enabled with "smtps" anyway).

nicolas-grekas commented:
Thank you @ssddanbrown.
